compiler/rustc_ast_lowering/src/lib.rs

```diff
@@ -227,6 +227,12 @@ impl ResolverAstLoweringExt for ResolverAstLowering {
     }
 }
 
+#[derive(Clone, Copy)]
+enum EffectContext {
+    Parametrized { host_param_id: LocalDefId },
+    Invariant { value: bool },
```

Comment:
I struggled a bit with the modelling of this variant. I know that `Invariant { value: true }` is super useless and that I should probably make it into a unit variant. However, I don't know what I should name it. I don't know if `EffectContext` should mention `const`-specific stuff, I opted against it. Otherwise, `AlwaysConst` would be the most obvious choice.